name.abuchen.portfolio/src/name/abuchen/portfolio/model/Bookmark.java

```diff
@@ -42,11 +48,63 @@ public boolean isSeparator()
 
     public String constructURL(Security security)
     {
-        String url = pattern.replace("{tickerSymbol}", encode(security.getTickerSymbol())); //$NON-NLS-1$
-        url = url.replace("{isin}", encode(security.getIsin())); //$NON-NLS-1$
-        url = url.replace("{wkn}", encode(security.getWkn())); //$NON-NLS-1$
-        url = url.replace("{name}", encode(security.getName())); //$NON-NLS-1$
+        boolean replacementDone = Boolean.FALSE;
+        String url = pattern;
+        Class<? extends Security> securityClass = security.getClass();
+        HashMap<String, String> types = new HashMap<String, String>();
+        TreeMap <Integer,Entry<String, String>> replaceTypes = new TreeMap<Integer,Entry<String, String>>();
+
+        
+        
+        types.put("{tickerSymbol}", "getTickerSymbol");
+        types.put("{isin}",  "getIsin");
+        types.put("{wkn}", "getWkn");
+        types.put("{name}",  "getName");
+
+       
+        for(Entry<String, String> type : types.entrySet()){
+            if(pattern.contains(type.getKey()))
+              replaceTypes.put(pattern.indexOf(type.getKey()), type);
+        }
 
+        for( Integer key :replaceTypes.keySet()){
+            Entry<String, String> replacement = replaceTypes.get(key);
+            Method method;
+            try
+            {
+                method = securityClass.getMethod(replacement.getValue());
+                String queryString = (String) method.invoke(security);
+
+                if(!replacementDone && queryString!= null && queryString.length() > 0 ){
+                        url = url.replace(replacement.getKey(), encode(queryString));
+                        replacementDone = Boolean.TRUE;
+                }
+                else 
+                    url = url.replace(replacement.getKey(), encode(""));
+                
+        
+            }
+            catch (NoSuchMethodException | SecurityException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e)
+            {
+                e.printStackTrace();
```

Comment:
Die Exceptions würde ich in RuntimeExcepition verpacke und weiterwerfen. Ich weiß in dem "name.abuchen.portfolio" Bundle kann man nicht einfach Fehler loggen. Aber die StackTrace mit printStackTrace landen quasi im Nirwana (der err output stream wird in eine Datei in den Configuration Folder umgeleitet). Dann lieber werfen und das UI Modul loggt einen Fehler.

Und: vermutlich fallen die alle weg wenn man die Reflection ausbaut.